### PR TITLE
Bluetooth: controller: Fix ISO broadcaster pre-transmission groups > 1

### PR DESCRIPTION
**Pull request by mtpr-ot**
For pre-transmission groups > 1, the broadcaster link layer would fetch incorrect payloads from the TX node queue.

Update payload index calculation to reference correct TX payload. Fix PTO FIXMEs and range checks.

bsim bap_broadcast_audio_source/sink tested with PTO=2 PTO Group Count = 2:
[bsim_bap_broadcast_source_sink_pto_2_ptogc_2.zip](https://github.com/user-attachments/files/18363537/bsim_bap_broadcast_source_sink_pto_2_ptogc_2.zip)

### CODE REVIEW

**Comment by cvinayak** [comment]
@mtpr-ot This PR was never put to ready for review, intentional? Seems this is valid change request, if there is no objection, I will put this for review and follow up (provided can do some testing if possible)

**Comment by Tronil** [comment]
Hi @cvinayak - Morten recently left Oticon, so I'm not sure if he is still active with this account.

In any case, we are using this change downstream, so I think it is just an oversight that this PR never left draft.

**Comment by cvinayak** [comment]
> Hi @cvinayak - Morten recently left Oticon, so I'm not sure if he is still active with this account.
> 
> In any case, we are using this change downstream, so I think it is just an oversight that this PR never left draft.

Ok, I will follow up on the failing bsim CI test, suspect an overlapping/drifting ACL role leading to missing ISO SDU reception...

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c [hunk 1]
```diff
@@ -627,9 +627,24 @@ static void isr_tx_common(void *param,
...
+
+			/* Calculate group index and multiplier for deriving
+			 * pre-transmission payload index.
+			 */
+			ptx_group_idx = ptx_idx / lll->bn;
+			ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn;
+			ptx_group_mult = (ptx_group_idx + 1) * lll->pto;
+			payload_index = ptx_payload_idx + ptx_group_mult * lll->bn;
+		} else {
+			payload_index  = lll->bn_curr - 1U;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:644]:
```suggestion
			payload_index = lll->bn_curr - 1U;
```

Any risk of underflow here?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c [hunk 2]
```diff
@@ -627,9 +627,24 @@ static void isr_tx_common(void *param,
...
+		}
+
+		payload_count = lll->payload_count - lll->bn + payload_index;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:647]: Similarly can `lll->bn` be larger than `lll->payload_count` here?

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 3]
```diff
@@ -1155,17 +1162,39 @@ static uint8_t ptc_calc(const struct lll_adv_iso *lll, uint32_t event_spacing,
...
 
 		/* Possible maximum Pre-transmission Subevents per BIS */
 		ptc = ((event_spacing_max - event_spacing) /
 		       (lll->sub_interval * lll->bn * lll->num_bis)) *
 		      lll->bn;
 
-		/* FIXME: Here we restrict to a maximum of BN Pre-Transmission
-		 * subevents per BIS
+		/* Required NSE */
+		nse = lll->bn * lll->irc;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:1174]: Ditto here; risk of overflow when multiplying 2 uint8_ts?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c [hunk 4]
```diff
@@ -627,9 +627,24 @@ static void isr_tx_common(void *param,
...
+			uint32_t ptx_group_mult;
+			uint8_t ptx_group_idx;
+
+			/* Calculate group index and multiplier for deriving
+			 * pre-transmission payload index.
+			 */
+			ptx_group_idx = ptx_idx / lll->bn;
+			ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn;
+			ptx_group_mult = (ptx_group_idx + 1) * lll->pto;
+			payload_index = ptx_payload_idx + ptx_group_mult * lll->bn;
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c]: Fixed.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c (hunk 1 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c]: No, bn_curr range of 1 to 7.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c (hunk 2 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:648]: No, `lll->payload_count` was pre-incremented by `lll->bn`  here: https://github.com/zephyrproject-rtos/zephyr/blob/42faa9438f8b0ddb827718b3d20699f9bf88e0c0/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c#L206-L207

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 3 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:1174]: No, `bn` is 3 bits and `irc` is 4 bits.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c (hunk 2 above)
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:648]: If `lll->latency_prepare` then that comment doesn't hold up. Can `lll->latency_prepare == 0`?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c [hunk 5]
```diff
@@ -625,11 +625,27 @@ static void isr_tx_common(void *param,
...
+			uint16_t ptx_group_mult;
+			uint8_t ptx_group_idx;
+
+			/* Calculate group index and multiplier for deriving
+			 * pre-transmission payload index.
+			 */
+			ptx_group_idx = ptx_idx / lll->bn; /* 5 bits */
+			ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn; /* 8 bits */
+			ptx_group_mult = (ptx_group_idx + 1) * lll->pto; /* 9 bits */
+			payload_index = ptx_payload_idx + ptx_group_mult * lll->bn; /* 13 bits */
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:642]: The comments are somewhat helpful. Currently they are too vague (I know from context that they refer to the maximum size of the result, but that's not clear for someone else reading this later). 

As you may be aware, this is a general concern when I'm reviewing controller PRs, and in some cases there are potential issues caught and found (e.g. you've modified the type of `ptx_group_mult` and `payload_index` based on my review). 

The comments are OK assuming that the size of these values never change. A more future proof approach would be to have asserts ensuring that we don't have silent and unintentional under- or overflows.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c [hunk 6]
```diff
@@ -625,11 +625,27 @@ static void isr_tx_common(void *param,
...
 	/* Get ISO data PDU, not control subevent */
 	if (!pdu) {
-		uint8_t payload_index;
+		uint16_t payload_index;
 
-		payload_index = (lll->bn_curr - 1U) +
-				(lll->ptc_curr * lll->pto);
-		payload_count = lll->payload_count + payload_index - lll->bn;
+		if (lll->ptc_curr) {
+			uint8_t ptx_idx = lll->ptc_curr - 1; /* max. nse 5 bits */
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:631]: Since `ptc_curr` is `uint32_t ptc_curr:4;`, then shouldn't this be 
```suggestion
			uint8_t ptx_idx = lll->ptc_curr - 1; /* max. nse 4 bits */
```

Which then also affect the bits of several of the values below?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c (hunk 2 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:648]: `lll->latency_prepare` is greater than `1` when the `lll->payload_count` is incremented.

https://github.com/zephyrproject-rtos/zephyr/blob/966122045b9f555373db6b6eeed53cc0a84d76ea/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c#L117-L123

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c (hunk 6 above)
Comment (Tronil) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:631]: > Now I do not remember why `ptc` is 4 bits! ... it can be as much as (NSE - 1); 1 transmission followed by (NSE - 1) pre-transmissions.
> 
> @Tronil and @wopu-ot , are you testing the boundary conditions internally in your tests for this PR changes?

We do not have extensive test coverage on the broadcaster side, no

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c [hunk 7]
```diff
@@ -625,11 +625,27 @@ static void isr_tx_common(void *param,
...
+			uint8_t ptx_payload_idx;
+			uint16_t ptx_group_mult;
+			uint8_t ptx_group_idx;
+
+			/* Calculate group index and multiplier for deriving
+			 * pre-transmission payload index.
+			 */
+			ptx_group_idx = ptx_idx / lll->bn; /* 5 bits */
+			ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn; /* 8 bits */
```
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:640]: > hmmm... isnt  ptx_payload_idx = ptx_idx - ((ptx_idx / lll->bn) * lll->bn; always zero! ?

I though so too at first, but due to integer division it isn't. 

e.g. pts_index = 5 and lll->bn = 3 then
```
ptx_group_idx = ptx_idx / lll->bn; /* will be 1 */ 
ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn; /* will be 2 */
```

If `ptx_group_idx = ptx_idx / lll->bn;` was not integer divided and resulted in 5/3, then `ptx_payload_idx` would have become 0. 

I expected that was intentional, but I'm not sure

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c (hunk 6 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c]: @Tronil while testing this PR for PTO=2 and Group Count = 2, I discovered a bug in ull_sync_iso.c that subsequent BIS subevent reception failed; `ptc` was not assigned the correct value. This bug will not show up if only one BIS is synchronized to.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c [hunk 8]
```diff
@@ -625,10 +625,26 @@ static void isr_tx_common(void *param,
...
+			ptx_group_mult = (ptx_group_idx + 1U) * lll->pto; /* 9 bits */
+			payload_index = ptx_payload_idx + ptx_group_mult * lll->bn; /* 13 bits */
+			LL_ASSERT(payload_index <= UINT8_MAX);
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:643]: Something looks a bit here now; you have a comment stating `13 bits` but you also assert that it's less <= 8 bits

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c (hunk 6 above)
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c]: Shouldn't the comment be updated?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c [hunk 9]
```diff
@@ -1336,6 +1335,31 @@ static void isr_done(void *param)
...
+		ptx_group_idx = ptx_idx / lll->bn; /* 5 bits */
+		ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn; /* 8 bits */
+		ptx_group_mult = (ptx_group_idx + 1U) * lll->pto; /* 9 bits */
+		payload_index = ptx_payload_idx + ptx_group_mult * lll->bn; /* 13 bits */
+		LL_ASSERT(payload_index <= UINT8_MAX);
+	} else {
+		payload_index  = lll->bn_curr - 1U; /* 3 bits */
+	}
+
+	return payload_index & 0xFF;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:1360]: If you want to ensure that you return the correct type, you should do this:

```suggestion
	return (uint8_t)payload_index;
```

`0xFF` is an `int` so you are effectively implicitly casting the type to an `int` here, and you already have an assert for the correct value range above

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 10]
```diff
@@ -201,12 +212,11 @@ static uint8_t big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bi
 				return BT_HCI_ERR_INVALID_PARAM;
 			}
 
-			/* FIXME: PTO is currently limited to BN */
-			if (!IN_RANGE(pto, 0x00, bn /*0x0F*/)) {
+			if (!IN_RANGE(pto, 0x00, 0x0F)) {
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:215]: Coverity will complain about `IN_RANGE` on unsigned integers when the first value is 0x00. 

```suggestion
			if (pto > 0x0F)) {
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 11]
```diff
@@ -1155,19 +1173,51 @@ static uint8_t ptc_calc(const struct lll_adv_iso *lll, uint32_t event_spacing,
...
+		 * overflow.
+		 */
 		ptc = ((event_spacing_max - event_spacing) /
 		       (lll->sub_interval * lll->bn * lll->num_bis)) *
 		      lll->bn;
 
-		/* FIXME: Here we restrict to a maximum of BN Pre-Transmission
-		 * subevents per BIS
+		/* Required NSE */
+		nse = lll->bn * lll->irc; /* 3 bits + 4 bits */
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:1189]:
```suggestion
		nse = lll->bn * lll->irc; /* 3 bits * 4 bits */
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 12]
```diff
@@ -1155,19 +1173,51 @@ static uint8_t ptc_calc(const struct lll_adv_iso *lll, uint32_t event_spacing,
...
+
+		/* FIXME: Do not remember why ptc is 4 bits, it should be 5 bits as ptc is a
+		 *        running buffer offset related to nse. Fix ptc and ptc_curr definitions,
+		 *        until then lets have an assert check here.
 		 */
-		ptc = MIN(ptc, lll->bn);
+		LL_ASSERT(ptc <= BIT_MASK(4));
 
-		return ptc;
+		return ptc & BIT_MASK(5);
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:1220]: Is the `& BIT_MASK(5)` valuable when you already have the assert?

File: subsys/bluetooth/controller/ll_sw/ull_sync_iso.c [hunk 13]
```diff
@@ -483,13 +483,20 @@ void ull_sync_iso_setup(struct ll_sync_iso_set *sync_iso,
 	lll->sub_interval = PDU_BIG_INFO_SUB_INTERVAL_GET(bi);
 	lll->max_pdu = bi->max_pdu;
 	lll->pto = PDU_BIG_INFO_PTO_GET(bi);
+	lll->bis_spacing = PDU_BIG_INFO_SPACING_GET(bi);
+	lll->irc = PDU_BIG_INFO_IRC_GET(bi);
 	if (lll->pto) {
-		lll->ptc = lll->bn;
+		uint8_t nse;
+
+		nse = lll->irc * lll->bn; /* 4 bits + 3 bits */
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:491]:
```suggestion
		nse = lll->irc * lll->bn; /* 4 bits * 3 bits */
```

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c (hunk 9 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c]: Updated code. ISO Sync Receiver has no data type limitation that is applicable for ISO Broadcaster wrt buffer indices (runtime check for overflows are present).

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 10 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c]: Fixed.

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 11 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c]: Fixed.

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 12 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c]: agree, fixed.

File: subsys/bluetooth/controller/ll_sw/ull_sync_iso.c (hunk 13 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_sync_iso.c]: Fixed.

**Review by Thalley — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by cvinayak — APPROVED** [review]
I am a co-author is this PR, but do approve the initial commits from the original author.
